### PR TITLE
GH-35718: [Go][Parquet] Fix for null-only encoding panic

### LINKED ISSUE #35718: [Go] [Parquet] panic writing with DeltaBinaryPacked when column only has nulls
### Describe the bug, including details regarding any error messages, version, and platform.

The writer panics if DeltaBinaryPacked encoding is requested for a column that is all null. The following small change to an existing test case demonstrates the issue. 

In case it matters, I'm using `go version go1.20.4 darwin/amd64`
I ran into this issue using arrow v12.0.0, but this changed testcase fails using the current head of main.

```
diff --git a/go/parquet/pqarrow/encode_arrow_test.go b/go/parquet/pqarrow/encode_arrow_test.go
index 877f584f2..0b83ec696 100644
--- a/go/parquet/pqarrow/encode_arrow_test.go
+++ b/go/parquet/pqarrow/encode_arrow_test.go
@@ -380,6 +380,8 @@ func TestWriteEmptyLists(t *testing.T) {
 
        props := parquet.NewWriterProperties(
                parquet.WithVersion(parquet.V1_0),
+               parquet.WithDictionaryDefault(false),
+               parquet.WithEncoding(parquet.Encodings.DeltaBinaryPacked),
        )
        arrprops := pqarrow.DefaultWriterProps()
        var buf bytes.Buffer
```

### Component(s)

Go

### PR DESCRIPTION
**Pull request by MagicBoost**
### Rationale for this change

closes: #35718 

### What changes are included in this PR?

Fix painc writing with DeltaBinaryPacked or DeltaByteArray when column only has nulls

### Are these changes tested?

Yes

- add a test writing nulls to columns with DeltaBinaryPacked / DeltaByteArray / DeltaLengthByteArray encodings

### Are there any user-facing changes?

No

* Closes: #35718

### CODE REVIEW

File: go/parquet/pqarrow/encode_arrow_test.go [hunk 1]
```diff
@@ -427,6 +427,44 @@ func TestWriteEmptyLists(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestWriteAllNullsWithDeltaEncoding(t *testing.T) {
+	sc := arrow.NewSchema([]arrow.Field{
+		{Name: "f1", Type: arrow.PrimitiveTypes.Int64, Nullable: true},
+		{Name: "f2", Type: arrow.ListOf(arrow.FixedWidthTypes.Date32)},
```
Comment (mapleFU) [inline go/parquet/pqarrow/encode_arrow_test.go:433]: Hmm why List is tested here?

File: go/parquet/pqarrow/encode_arrow_test.go [hunk 2]
```diff
@@ -427,6 +427,44 @@ func TestWriteEmptyLists(t *testing.T) {
...
+
+	props := parquet.NewWriterProperties(
+		parquet.WithVersion(parquet.V1_0),
+		parquet.WithDictionaryDefault(false),
+		parquet.WithEncodingFor("f1", parquet.Encodings.DeltaBinaryPacked),
+		parquet.WithEncodingFor("f2", parquet.Encodings.DeltaBinaryPacked),
+		parquet.WithEncodingFor("f3", parquet.Encodings.DeltaByteArray),
+		parquet.WithEncodingFor("f4", parquet.Encodings.DeltaByteArray),
+		parquet.WithEncodingFor("f5", parquet.Encodings.DeltaLengthByteArray),
+		parquet.WithEncodingFor("f6", parquet.Encodings.DeltaLengthByteArray),
```
Comment (mapleFU) [inline go/parquet/pqarrow/encode_arrow_test.go:456]: Would you mind testing all these encoders?

Comment (mapleFU) [reply go/parquet/pqarrow/encode_arrow_test.go:456]: ( Maybe `go/parquet/internal/encoding/encoding_test.go` is also a good place for testing?)

**Comment by mapleFU** [comment]
(I'm afraid https://github.com/apache/arrow/pull/39347/files does the similiar things but less than your contributions in this patch, but it doesn't get merged, I will wait @zeroshade for a while...)

File: go/parquet/pqarrow/encode_arrow_test.go (hunk 1 above)
Comment (MagicBoost) [reply go/parquet/pqarrow/encode_arrow_test.go:433]: Strictly speaking this is not necessary，I just copy the previous test and want to add some simple tests on nested formats.

File: go/parquet/pqarrow/encode_arrow_test.go (hunk 2 above)
Comment (MagicBoost) [reply go/parquet/pqarrow/encode_arrow_test.go:456]: I don't get it, I have already used these encoders testing nulls encoding.

Comment (MagicBoost) [reply go/parquet/pqarrow/encode_arrow_test.go:456]: There is no panic or error happens in the encoding process, panic happens when writing data into a file (call written function on a nil bitWriter not inited).

Comment (mapleFU) [reply go/parquet/pqarrow/encode_arrow_test.go:456]: In this patch we check DeltaBinaryPack and DeltaByteArray. DeltaLengthByteArray uses DeltaBinaryPack, and DeltaByteArray uses DeltaLengthByteArray. So I think it's ok to testing them

Besides, there're some other encodings: RleBoolean, Plain ..etc. Should we also testing them?

Comment (MagicBoost) [reply go/parquet/pqarrow/encode_arrow_test.go:468]: Ok, I added more encording tests.

File: go/parquet/pqarrow/encode_arrow_test.go (hunk 1 above)
Comment (zeroshade) [reply go/parquet/pqarrow/encode_arrow_test.go:433]: It's good to ensure we're testing nulls for multiple types

**Comment by zeroshade** [comment]
I've rebased this and fixed the conflict after merging my change. Shrinks this down to a fairly small change plus all the new tests. This all seems good to me assuming that all the CI passes. Thanks much @MagicBoost !

**Comment by mapleFU** [comment]
```go
// EstimatedDataEncodedSize returns the current number of bytes that have
// been buffered so far
func (enc *PlainBooleanEncoder) EstimatedDataEncodedSize() int64 {
	return int64(enc.sink.Len() + int(bitutil.BytesForBits(int64(enc.wr.Pos()))))
}
```

@zeroshade This might suffer from problem when `enc.wr == nil`. However, if all input is null, the `Put` would also be called here...Would we leave it here or also fixing this?

**Pull request merged**